fs/readers.go

```diff
@@ -0,0 +1,135 @@
+package fs
+
+import (
+	"errors"
+	"io"
+)
+
+// A ChunkReader implements the io.Reader interface to only serve a chunk of the data
+// It will allow read until specific size of from the internal Reader hiding it's full size
+// Since ChunkReader will only Read data from internal Reader when requested it will
+// play nicely with the Account to reflect current speed
+type ChunkReader struct {
```

Comment:
I was not aware of the `io.LimitedReader` I was trying to find something similar but I guess I missed it. I didn't want to change the current implementation of the fs too much but I think we caneasily change it so there won't be need to the Len method which will make the entire implementation.

As with the accurate speed. What I found is that the core of the issue is reading the full chunk into the a buffer with `io.ReadFull` this will cause the account to think the entire chunk has been uploaded even though the upload event started yet. when the parts are being uploaded they will not count anymore since they are being read from the buffer which does not affect the account any more. 
I tried looking at the commit you mentioned and use the `fs.AccountPart` method before implementing this but I found that it had the same issue (But maybe I used it wrong).

I think that if you are using a fast connection this issue is harder to notice cause if the chunk will be finished uploading by the time you get the stats they will be fine, but if you have a slow connection where a chunk might take a few minutes to upload the first stats will look amazingly fast (much faster than my connection will ever allow) and the following stats will show the speed declining rapidly